docs(gatsby): Query Filters (#24775)
gatsbyjs/gatsby: docs/docs/query-filters.md

---
title: Query Filters with GraphQL in Gatsby
---

> This documentation is up to date as of `gatsby@2.23.0`.

## Summary

> Gatsby stores all data loaded during the source-nodes phase in Redux and it allows you to write GraphQL queries to query that data. This data, stored as individual "nodes", can be searched through using a query language that is a subset of [MongoDb queries](https://docs.mongodb.com/manual/reference/operator/query/).

Reviewer: I'd rather say that we are inspired by mongodb rather than claiming we implement a subset.

> When you write a GraphQL query, the `filter` parts will be processed and Gatsby will return all the nodes that match each of the given filters. GraphQL does the rest.

Reviewer: ```suggestion
Filtering is used in GraphQL root fields of Node types (e.g. for File type it would be `file` and `allFile`). `filter` GraphQL argument is passed to the filtering system and will return all the nodes that match each of the given filters. The resto of the processing, such as pagination, is handled on GraphQL resolver level.
```

Unfortunately Sift did not align with how Gatsby used it and so a custom system was written to slowly replace it. This system was called "fast filters" and as of gatsby@2.23.0 (June 2020) the Sift library is no longer used.

## Query Language

The syntax and API used by the filters is based on [the MongoDB query syntax](https://docs.mongodb.com/manual/reference/operator/query/) but keep in mind only a subset of comparators is supported in Gatsby.

In general a filter has a "filter path" component ending with a "comparator", and a "filter value". It's a way of programmatically asking "show me all nodes where `node[i].a.b.c` leads to a value less than `5`".

- path: an exact property path when traversing an object (`{ a: { b: { c: { op: x } } } }` has path `a`, `a.b`, and `a.b.c`)
- value: the needle to compare against the value at the end of a path
- comparator: a way to tell the system how to compare the given filter value to the resulting value at the end of the path. Should it be equal, lower than, or match in a special way?

> Filters are generally applied to a subset of all nodes, matching a specific `node.internal.type` value, like `` `PageData` `` or `` `markdownRemark` ``.

Author: How do you usually handle a quoted string in a code span?

We use backticks for strings as per our styleguide. To have code with backticks you start and end with more backticks. So that's why it's there.

> Filters are generally applied to a subset of all nodes, matching a specific `node.internal.type` value, like `` `PageData` `` or `` `markdownRemark` ``.

Reviewer 2: on other places in the doc you use normal ones 🤔 
```markdown
`'name'`
```

When applying a filter, the path will be checked for each node and the resulting value is compared against the requested filter value with the rules of the comparator used.

## Example

Let's say we have a filter and a set of nodes:

```js
filter = { post: { author: { name: { eq: "Alex" } } } }

nodes = [
  { id: 1, post: { title: "Hello, world!", author: { name: "Alex" } } },
  { id: 2, post: { title: "Debugging Gatsby", author: { name: "Clarissa" } } },
  { id: 3, post: { title: "Publishing on Gatsby", author: { name: "Ika" } } },
  { id: 4, post: { title: "Fixed a bug", author: { name: "Alex" } } },
]
```

The filter path is `post.author.name`, the comparator is `eq`, and the filter value is `'Alex'`.

This ought to return two nodes, the ones with id 1 and 3, because those are the only ones where the result value `node[i].post.author.name` equals `'Alex'`.

A filter path can combine multiple paths, for example:

- `{ id: { gt: 2 }, post: { author: { name: { eq: 'Alex' } } } } }` would return only the node with id 4
- `{ post: { title: { regex: '/Gatsby/i' }, author: { eq: 'Ika' } } }` would return the node with id 3.

### Supported comparators

> Gatsby supports a subset of MongoDB comparators, namely, op matches every node where the result value;

Author: Yeah this can be phrased better.

  - Can only be used with numbers, strings, booleans, and null
  - If filter value is `null`;
    - `lt` and `gt` will never match anything
    - `lte` and `gte` will only match nodes with a result value of exactly `null`
  - Will never return nodes with partial paths
  - Weak comparison due to the nature of the JavaScript operators
- `regex`, `glob`
  - This is like `new RegExp(filterValue).test(resultValue)` (with caveats for the filterValue syntax)
  - Glob pattern is converted to a JavaScript RegExp with [micromatch](https://github.com/micromatch/micromatch)
  - The `regex` filterValue must be a stringified regular expression, including leading and trailing forward slash and optional flags; Like `"/foo/g"`
  - Never returns nodes with partial paths
  - While testing, result values are explicitly casted to a string through `String(resultValue)` before passing it to `regex.test()`

### Nulls and partial paths

Due to legacy support for MongoDB compatibility, there are edge cases for each comparator when it comes to `null` values and for partial or non-existing paths. It's best to try and avoid these cases altogether.

## Performance

> The key metric that impacts the performance of your queries is the node count for your type of nodes. You can see a dump of these counts when using the `--verbose` flag while building (`gatsby build --verbose`). It tells you the node counts per type during bootstrap. There will be a separate message of the number of page nodes, since they are generated later.

Reviewer 3: ```suggestion
The key metric that impacts the performance of your queries is the node count for your type of nodes. You can see a dump of these counts when using the `--verbose` flag while building (`gatsby build --verbose`). It tells you the node counts per type during the bootstrap sequence. There will be a separate message of the number of page nodes, since they are generated later.
```

While the number of pages is definitely a factor for the number of nodes, there can be other factors at play that cause many internal nodes even when the actual number of pages seems to be low.

The actual performance of filtering is a combination of the number of nodes you have to search through, the kind of comparator being used, and how many unique filters a query consists of.

For a low node count none of this matters. Roughly speaking, a site with fewer than 1000 - 10000 nodes should not have to worry too much about this.

> When scaling up these are some rules of thumb to keep in mind when filtering:

Reviewer 3: "Rules of thumb" is one of those phrases we should try to avoid if possible, open to other suggestions.

```suggestion
When scaling up these are some rules to keep in mind when filtering:
```

Author: Ok

---
title: Query Filters with GraphQL in Gatsby
---

> This documentation is up to date as of `gatsby@2.23.0`.

## Summary

Gatsby stores all data loaded during the source-nodes phase in Redux and it allows you to write GraphQL queries to query that data. This data, stored as individual "nodes", can be searched through using a query language that is inspired by [MongoDb queries](https://docs.mongodb.com/manual/reference/operator/query/).

Filtering is used in GraphQL root fields of Node types (e.g. for File type it would be `file` and `allFile`). `filter` GraphQL argument is passed to the filtering system and will return all the nodes that match each of the given filters. The resto of the processing, such as pagination, is handled on GraphQL resolver level.

> Filtering is used in GraphQL root fields of Node types (e.g. for File type it would be `file` and `allFile`). The GraphQL `filter` argument is passed to the filtering system and will return all the nodes that match each of the given filters. The rest of the processing, such as pagination, is handled on GraphQL resolver level.

Reviewer 2: double content: this paragraph is the same as before?

Reviewer 2: already fixed in #26371 `Fix typos and linter dictionary errors`
